Feature: 포스트 변경 API 및 테스트 구현

yooooonshine opened the pull request:
## 개요
- close #14

## 작업사항
### 이미지 도용 공격 방어
![image](https://github.com/user-attachments/assets/5b0935f9-871a-4123-92d7-046006b6a32a)
기존에 포스트를 생성할 때 ImageID를 입력하게 했고, 이 때 해당 Image에 postId를 외래키로 넣어주었다.
다만 다른 포스트에서 기존 포스트 이미지 ID를 적으면 PostId가 변경되어 도용되는 일이 발생했다.

![image](https://github.com/user-attachments/assets/ca5f7430-e339-45e0-aa0a-1399cc689144)
이로 인해 한번 결정된 PostId는 수정하지 못하도록 변경했다.

### 포스트 변경 로직 및 테스트 구현
포스트 변경 로직 시 이미지 처리가 가장 중요했다.

![image](https://github.com/user-attachments/assets/804d3eca-1626-4b9f-9380-7c9cc612ebe7)
없어진 Image들은 삭제하게 했으며, 추가된 이미지에 대해서는 Image에 외래키를 등록하도록 했다.

이에 맞게
- 포스트에서 이미지를 제외했을 때 삭제되는 지 테스트
- 이미지를 도용하려 했을 때 에러 테스트 등을 작성했다.

- 그 외에 Controller, Integration, Service 테스트도 작성했다.

yooooonshine committed:
feat: 포스트 변경 로직 구현

- 기존 이미지들은 남기고, 없어진 이미지들은 삭제, 새로운 이미지들은 저장하게 끔 구현

yooooonshine committed:
feat: 이미지 도용 방지 기능 도입

- 기존에는 Post 생성 및 업데이트 시, 이미 다른 Post에서 사용중인 Image도 뺏어올 수 있었다. 이를 개선했다.

yooooonshine committed:
Merge branch 'develop' of https://github.com/Modic-2025/modic_backend into feature/14-update-article

yooooonshine committed:
fix : PostId가 존재하지만 같은 PostId일 때도 예외가 발생하는 에러 해결

- 특정 게시물에 속한 postId에 대해 같은 postId를 넣는 것은 허용해야 한다.

yooooonshine committed:
feat : 이미지 변경 API 테스트 코드 작성

- Unit 테스트
- controller 테스트
- 통합 테스트

goalSetter09 commented on src/main/java/hanium/modic/backend/domain/post/repository/PostImageEntityRepository.java:
```diff
@@ -3,9 +3,13 @@
 	boolean existsByImagePath(String imagePath);
+
+	@Query("SELECT pi FROM PostImageEntity pi WHERE pi.id IN :ids")
+	List<PostImageEntity> findAllByIds(List<Long> ids);
```
`findAllByIdIn`를 사용하면 쿼리 문을 직접 작성할 필요가 없지 않나요??
직접 쿼리 문을 작성하신 이유가 있으신가요?

yooooonshine replied:
만약 in 쿼리가 아니라 jpa 메서드를 사용하면 findById로 id 개수만큼 쿼리를 날려야했기에 한번에 가져오고자 작성했습니당!

goalSetter09 replied:
```
@Query("SELECT pi FROM PostImageEntity pi WHERE pi.id IN :ids")
List<PostImageEntity> findAllByIds(List<Long> ids);
```

```
List<PostImageEntity> findAllByIdIn(List<Long> ids);
```
두 코드가 동일한 쿼리 날리는걸로 알고 있는데 아닌가요??

goalSetter09 commented on src/main/java/hanium/modic/backend/web/post/controller/PostController.java:
```diff
@@ -68,4 +70,14 @@ public ResponseEntity<ApiResponse<Void>> deletePost(@PathVariable Long id) {
+		@PathVariable Long id,
+		@RequestBody @Valid UpdatePostRequest request) {
+		postService.updatePost(id, request.title(), request.description(), request.commercialPrice(),
+			request.nonCommercialPrice(), request.imageIds());
```
이미지 id를 클라이언트에서 알고 있나요?
게시글 상세 조회 api에서 응답이 iamgeUrl을 리스트로 되어 있는데, 혹시 요청이 이걸로 가능한건지 궁금합니다.

yooooonshine replied:
오,,,그러네요.. 이에 맞게 게시글 조회도 수정하겠습니다!!

goalSetter09 commented on src/main/java/hanium/modic/backend/web/post/dto/request/UpdatePostRequest.java:
```diff
@@ -0,0 +1,32 @@
+	@Length(min = 1, max = 20)
+	String title,
+
+	String description,
```
description은 없어도 됐나요?!

goalSetter09 commented on src/main/java/hanium/modic/backend/web/post/controller/PostController.java:
```diff
@@ -68,4 +70,14 @@ public ResponseEntity<ApiResponse<Void>> deletePost(@PathVariable Long id) {
 		return ResponseEntity.status(NO_CONTENT).body(ApiResponse.noContent());
 	}
+
+	@PatchMapping("/{id}")
```
patch 메서드의 경우, 변경할 필드만 request로 받아서 수정하는 것으로 알고 있는데,
현재 `UpdatePostRequest`의 경우 모든 필드를 받아서 서비스로 넘기고 있어서 PUT을 사용하시거나, 아니면 수정할 필드만 입력받도록 변경하시는 건 어떨지 제안드립니다!

yooooonshine replied:
id를 변경하지 않아 patch로 했는데 생각해보니 put이 맞아보입니다! PUT으로 수정하겠습니다!!

goalSetter09 commented on src/main/java/hanium/modic/backend/domain/post/service/PostService.java:
```diff
@@ -103,4 +103,34 @@ public void deletePost(Long postId) {
+		// imageIds에 포함되지 않은 이미지 삭제
+		postImages.stream()
+			.filter(postImageEntity -> !imageIds.contains(postImageEntity.getId()))
+			.forEach(postImageEntity -> postImageService.deleteImage(postImageEntity.getId()));
```
imageUtil에서 S3의 deleteObjects()를 활용하는 deleteImages를 새로 만들고,
이미지의 id를 모아서 한 번에 요청하면 성능상으로 더 좋을 것 같다는 생각이 드는데 어떻게 생각하시나요?!
지금은 매 요청마다 s3client 요청이 일어나서 한 번의 api 요청마다 최대 8개(이미지 최대 개수)의 s3client 콜이 있을 수 있을 것 같아요!

yooooonshine replied:
오 좋은 생각인 거 같습니다! 한번 시도해보겠습니다!

goalSetter09 reviewed (commented):
수고하셨습니다!

yooooonshine committed:
feat : 여러 이미지 한 요청으로 삭제하게 변경 및 포스트 업데이트 patch에서 put으로 변경

- 포스트 업데이트 patch -> put으로 바꿔 전체를 바꾼다는 느낌을 강조했다.
- S3에 여러 이미지 삭제시 기존에는 여러 요청을 보냈는데 한 요청으로 처리하도록 개선했다.
- S3 변경에 따라 테스트도 변경했다.

yooooonshine committed:
feat : 불필요한 게시글 삭제 controller 테스트 삭제

- id는 Long으로 받는데 이에 String을 전달하면 ArgumentResolver에서 BindException이 발생하기에 굳이 서비스 단에서 에러를 잡을 필요가 없어보였다.
- -1이라는 값 또한 삭제 요청이 어쩌피 없는 것을 삭제하려는 것이기 POST_NOT_FOUND_EXCEPTION이 발생하므로 잘못된 파라미터라고 볼 수 없다.

yooooonshine committed:
refactor : 게시글 조회 기능

yooooonshine reviewed (commented):
감사합니다!

yooooonshine committed:
feature: description validation 추가
